### DISCUSSION TITLE
Rebuilding the shared libc++ from NDK

### DISCUSSION THREAD

**Discussion by 0Xellos** [android/ndk#1696] [Q&A]
[sym_ours.txt](https://github.com/android/ndk/files/8449749/sym_ours.txt)
[sym_theirs.txt](https://github.com/android/ndk/files/8449750/sym_theirs.txt)
For simplicity, let's talk about NDK r22b and armv7 architecture. I'm on Ubuntu 20.04 (x86_64).

I'm trying to reproduce the exact `libc++_shared.so` for armv7 that's found in the NDK package. **What I've tried so far:**

1. Followed the official documentation https://android.googlesource.com/platform/ndk/+/ndk-release-r22/docs/Building.md to build NDK in docker. I had to kill `repo sync` after it drained 30 GB of disk space without any sign of stopping. When I tried normal `git clone` on branch `ndk-release-r22`, it worked but docker build failed because of `bootstrap` module in Python.

2. Take the conanfile from https://github.com/android/ndk/issues/1534#issuecomment-883201516, remove lines 13, 18-20, 47-49 and 55-63. The build progressed to linking step, where it crashed because of undefined symbols. Move to the build directory `$HOME/.conan/data/AndroidNdk/r22b/_/_/build/44fcf6b9a7fb86b2586303e3db40189d3b511830/armeabi-v7a`, copy-paste the link command, delete `-nodefaultlibs` flag, replace shared `-lc++abi` by static `lib/libc++abi.a and run the link command manually:
```
$HOME/.conan/data/AndroidNdk/r22b/_/_/source/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=armv7-none-linux-androideabi16 --gcc-toolchain=$HOME/.conan/data/AndroidNdk/r22b/_/_/source/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=$HOME/.conan/data/AndroidNdk/r22b/_/_/source/android-ndk-r22b/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security   -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers  -Wno-l
[sym_ours.txt](https://github.com/android/ndk/files/8449739/sym_ours.txt)
[sym_theirs.txt](https://github.com/android/ndk/files/8449740/sym_theirs.txt)
ong-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections -Oz  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libgcc_real.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--fatal-warnings -Wl,--exclude-libs,libunwind.a -Wl,--no-undefined -Qunused-arguments  -Wl,-z,defs -Wl,--color-diagnostics -shared -Wl,-soname,libc++_shared.so -o lib/libc++_shared.so projects/libcxx/src/CMakeFiles/cxx_shared.dir/algorithm.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/any.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/atomic.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/barrier.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/bind.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/charconv.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/chrono.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/condition_variable.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/condition_variable_destructor.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/debug.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/exception.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/functional.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/future.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/hash.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/ios.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/iostream.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/locale.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/memory.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/mutex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/mutex_destructor.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/new.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/optional.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/random.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/regex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/shared_mutex.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/stdexcept.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/string.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/strstream.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/system_error.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/thread.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/typeinfo.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/utility.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/valarray.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/variant.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/vector.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/operations.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/directory_iterator.cpp.o projects/libcxx/src/CMakeFiles/cxx_shared.dir/filesystem/int128_builtins.cpp.o  -nodefaultlibs  -lc  -landroid_support  lib/libc++abi.so  -latomic -lm
```

I concluded that I'll be using conan+cmake since that allowed me to actually build a libcxx library and I'm using them in other projects.

**The question:** How do I change my build so that I'd get the exact `libc++_shared.so` from the NDK? There are small differences at this point.

Let's start with the list of symbols as a kind of checksum. I threw away symbol values and types and some symbols I'm not interested in.
```
nm lib/libc++_shared.so -C | sed -e 's/^[0-9a-f]* \s*[a-zA-Z] //g' -e '/^\$[adt].*$/d' -e '/GCC_except_table/d' | sort > sym_ours.txt
nm ../../../source/android-ndk-r22b/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a/libc++_shared.so -C | sed -e 's/^[0-9a-f]* \s*[a-zA-Z] //g' -e '/^\$[adt].*$/d' -e '/GCC_except_table/d' | sort > sym_theirs.txt
```
Resulting files are attached. The difference is very small, just over 100 lines even though both files have over 4000. That tells me I'm on the right path. First off, how do I need to modify my build to get identical files?

**Comment by DanAlbert** [comment]
You were right with step one when you tried following the docs. You just have to wait for sync to finish.

**Reply by 0Xellos** [reply]
Let's just accept that it's Not An Option(tm) for me. I made some cmake build, so now I'm looking for the right cmake build.

**Reply by enh-google** [reply]
then you can't achieve your goal of reproducing our build exactly, and we can't help you.

looks like you need about 80GiB for a _built_ NDK tree; 70 GiB for just the source. (this tree is a bit old, but should be close enough.)
```
~/ndk$ du -sh .
74G	.
~/ndk$ rm -rf out
~/ndk$ du -sh .
67G	.
```

**Reply by 0Xellos** [reply]
Why? I don't believe it's impossible to make the same build if it's possible to make a similar build. Telling me to give up isn't helpful.

**Reply by enh-google** [reply]
i'm not telling you to give up --- i'm telling you we won't be able to help, and have no idea how much work it will be.

**Reply by DanAlbert** [reply]
I can more or less guarantee it'll be more work than following the docs though.

**Reply by pirama-arumuga-nainar** [reply]
`--depth=1` or `--partial-clone` to `repo init` will reduce the size of the checked-out tree.
